src/server.rs

```diff
@@ -817,82 +850,175 @@ where
+                        metadata(&path2)
                         .map(|attr| FileTime::from_last_modification_time(&attr))
-                    {
-                        Ok(mtime) => Some((archive, mtime)),
-                        _ => None,
+                        .ok()
+                        .map(move |filetime| {
+                            (path2.clone(),filetime)
+                        })
+                    }
+                };
+                f_ok(opt)
+            });
+
+        let lookup_compiler =
+            w_fallback.and_then(move |opt : Option<(PathBuf, FileTime)>| {
+                let (resolved_compiler_path, mtime)
+                    = opt.expect("Must contain sane data, otherwise mtime is not avail");
+
+
+                let dist_info = match me1.dist_client.get_client() {
+                    Ok(Some(ref client)) => {
+                        if let Some(archive) = client.get_custom_toolchain(&resolved_compiler_path) {
+                            match metadata(&archive)
+                                .map(|attr| FileTime::from_last_modification_time(&attr))
+                            {
+                                Ok(mtime) => Some((archive, mtime)),
+                                _ => None,
+                            }
+                        } else {
+                            None
+                        }
+                    }
+                    _ => None,
+                };
+
+                let opt = match me1.compilers.borrow().get(&resolved_compiler_path) {
+                    // It's a hit only if the mtime and dist archive data matches.
+                    Some(&Some(ref entry)) => {
+                        if entry.mtime == mtime && entry.dist_info == dist_info {
+                            Some(entry.compiler.clone())
+                        } else {
+                            None
+                        }
+                    }
+                    _ => None,
+                };
+                f_ok((resolved_compiler_path, mtime, opt, dist_info))
+            });
+
+        let obtain =
+            lookup_compiler.and_then(move |(resolved_compiler_path, mtime, opt, dist_info) : (PathBuf, FileTime, Option<Box<dyn Compiler<C>>>, Option<(PathBuf,FileTime)> )| {
+                match opt {
+                    Some(info) => {
+                        trace!("compiler_info cache hit");
+                        f_ok(Ok(info))
+                    }
+                    None => {
+                        trace!("compiler_info cache miss");
+                        // Check the compiler type and return the result when
+                        // finished. This generally involves invoking the compiler,
+                        // so do it asynchronously.
+
+                        // this must be done with the passed `path`, since that is the proxy
+                        // only this will work in all cases
```

Comment:
I'm confused about this comment.  What is it trying to explain?

Reply:
True, clarification pending..